### DISCUSSION TITLE
History Graph - Callback (v3.1.0-alpha 52)

### DISCUSSION THREAD

**Discussion by hosabeni** [Zefau/ioBroker.jarvis#1939] [Q&A]
Ich versuche gerade über die Callback Funktion des History Graph  Widgets die angezeigten Werte in der Zeit zu verschieben.
Leider wirkt die Eingabe von `ts => ts = ts - 10000` auf den Wert `val` und nicht auf `ts`. Die angezeigten Werte werden negativ.
Ist meine Eingabe falsch oder ist das ein Fehler im History Graph Widget?

![grafik](https://user-images.githubusercontent.com/77939385/193468660-c92fa8dd-5b2a-4b63-80b0-b26e12f5e6c9.png)

**Comment by Zefau** [comment]
Der Callback dient der Änderung der Werte, nicht der Zeit. Du kannst auf der X-Achse das Label anpassen, siehe https://echarts.apache.org/en/option.html#xAxis.axisLabel.formatter

**Reply by hosabeni** [reply]
> Der Callback dient der Änderung der Werte, nicht der Zeit. Du kannst auf der X-Achse das Label anpassen, siehe https://echarts.apache.org/en/option.html#xAxis.axisLabel.formatter

Ich hätte hier noch die Frage, wie ich den Code in den JSON packe damit er auch ausgeführt wird?

```
function (value, index) {
    var date = new Date(value - 10000);
    var month = date.getMonth() + 1;
    return month ;
}
```
Ich habe hier ein Beispiel aus deinem Link als Vorlage verwendet.

**Reply by Zefau** [reply]
```
(value, ts, index) => {
    var date = new Date(value - 10000);
    var month = date.getMonth() + 1;
    return month ;
}
```

**Reply by hosabeni** [reply]
Wenn ich denn Code unter Callback eingebe wird der Wert `val` verändert. Ich will aber `ts` verändern bzw. was als Datum/ Monat der angezeigt wird verschieben/ anpassen. 
Der Grund dafür ist, dass der Adapter SourceAnalytix die Werte erst einige ms nach 0Uhr aktualisiert und dann die Werte mit dem falschen Monat in der Datenbank stehen. Was bedeutet, das der falsche Monat unter dem Balken steht. Dies würde das gerne mit dieser Funktion korrigieren wie es hier [https://echarts.apache.org/en/option.html#xAxis.axisLabel.formatter](https://echarts.apache.org/en/option.html#xAxis.axisLabel.formatter) unter `formatter` mit einer Funktion möglich ist. 

![grafik](https://user-images.githubusercontent.com/77939385/193665478-092e4213-4476-4fb8-9fda-dc5a11043627.png)

**Reply by Zefau** [reply]
Du hast die Funktion in der Konfiguration unter X Achse drin?

`{ "formatter": "...." }` ?

**Reply by hosabeni** [reply]
Wenn ich die Funktion so in die X Achse Konfiguration eintrage:

![grafik](https://user-images.githubusercontent.com/77939385/193895167-8a3c08db-904a-4213-afd7-598c568c7c73.png)

Erhalte ich das (die Funktion wird als Text angezeigt):

![grafik](https://user-images.githubusercontent.com/77939385/193895326-fa9162ac-212d-4c5f-b327-93d9bceab1fe.png)

**Reply by hosabeni** [reply]
@Zefau Hast du noch eine Idee was ich machen bzw. testen könnte?

**Comment by hosabeni** [comment]
Dann ist die Vorgabe und der Infotext da an der Stelle etwas verwirrend.


![grafik](https://user-images.githubusercontent.com/77939385/193471912-54470624-d054-4ed1-baae-8aa1410f2632.png)

**Comment by Zefau** [comment]
Das sind die Übergabewerte. Du kannst dort eine Funktion definieren, die `val`, `ts` und `index` als Parameter nutzt (der Teil links von `=>`). Der Rückgabewert ist aber `val` (der Teil rechts von `=>`). Die Notation sagt das quasi aus.
